lib/trie/trie_test.go

```diff
@@ -1,1479 +1,3081 @@
+					},
+				},
+			},
+			prefix: []byte{1, 0, 2},
+			limit:  1,
+			newParent: &node.Leaf{
+				Key:   []byte{1},
+				Value: []byte{1},
+				Dirty: true,
+			},
+			updated:    true,
+			allDeleted: true,
 		},
 	}
 
-	for _, testCase := range cases {
-		for _, prefix := range prefixes {
-			trieDelete := NewEmptyTrie()
-			trieClearPrefix := NewEmptyTrie()
-
-			for _, test := range testCase {
-				trieDelete.Put(test.key, test.value)
-				trieClearPrefix.Put(test.key, test.value)
-			}
+	for name, testCase := range testCases {
+		testCase := testCase
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
 
-			prefixedKeys := trieDelete.GetKeysWithPrefix(prefix)
-			for _, key := range prefixedKeys {
-				trieDelete.Delete(key)
-			}
+			trie := testCase.trie
 
-			trieClearPrefix.ClearPrefix(prefix)
+			newParent, updated, allDeleted := trie.clearPrefixLimit(testCase.parent,
+				testCase.prefix, &testCase.limit)
 
-			require.Equal(t, trieClearPrefix.MustHash(), trieDelete.MustHash(),
-				fmt.Sprintf("tries not equal! prefix=0x%x\n, %s, %s", prefix, trieClearPrefix, trieDelete),
-			)
-		}
+			assert.Equal(t, testCase.newParent, newParent)
+			assert.Equal(t, testCase.expectedLimit, testCase.limit)
+			assert.Equal(t, testCase.updated, updated)
+			assert.Equal(t, testCase.allDeleted, allDeleted)
+			assert.Equal(t, testCase.expectedTrie, trie)
+		})
 	}
 }
 
-func TestSnapshot(t *testing.T) {
-	tests := []Test{
-		{key: []byte{0x01, 0x35}, value: []byte("spaghetti"), op: PUT},
-		{key: []byte{0x01, 0x35, 0x79}, value: []byte("gnocchi"), op: PUT},
-		{key: []byte{0x01, 0x35, 0x79, 0xab}, value: []byte("spaghetti"), op: PUT},
-		{key: []byte{0x01, 0x35, 0x79, 0xab, 0x9}, value: []byte("gnocchi"), op: PUT},
-		{key: []byte{0x07, 0x3a}, value: []byte("ramen"), op: PUT},
-		{key: []byte{0x07, 0x3b}, value: []byte("noodles"), op: PUT},
-		{key: []byte{0xf2}, value: []byte("pho"), op: PUT},
+func Test_Trie_deleteNodes(t *testing.T) {
+	t.Parallel()
+
+	testCases := map[string]struct {
+		trie         Trie
+		parent       Node
+		prefix       []byte
+		limit        uint32
+		newLimit     uint32
+		newNode      Node
+		oneDeletion  bool
+		expectedTrie Trie
+	}{
+		"zero limit": {
+			parent: &node.Leaf{
+				Key: []byte{1},
+			},
+			newNode: &node.Leaf{
+				Key: []byte{1},
+			},
+		},
+		"nil parent": {
+			limit:    1,
+			newLimit: 1,
+		},
+		"update leaf generation": {
+			trie: Trie{
+				generation: 1,
+			},
+			parent: &node.Leaf{},
+			newNode: &node.Leaf{
+				Generation: 1,
+			},
+			expectedTrie: Trie{
+				generation: 1,
+			},
+		},
+		"update branch generation": {
+			trie: Trie{
+				generation: 1,
+			},
+			parent: &node.Branch{},
+			newNode: &node.Branch{
+				Generation: 1,
+			},
+			expectedTrie: Trie{
+				generation: 1,
+			},
+		},
+		"delete leaf": {
+			parent:   &node.Leaf{},
+			limit:    2,
+			newLimit: 1,
+		},
+		"delete branch without value and leave limit unchanged": {
+			parent:   &node.Branch{},
```

Comment:
same comment as below

Reply:
Done in https://github.com/ChainSafe/gossamer/pull/2210/commits/bfde0a2e4e1ba9f097aebbb30235038b5086bbc1